docs/couler-tekton-design.md

```diff
@@ -0,0 +1,70 @@
+# Couler Tekton Design
+
+This document outlines the design of tekton backend for core Couler APIs.
+
+## Goals
+
+* Design the tekton backend implementation for core Couler APIs.
+
+## Non-Goals
+
+* Provide implementation details.
+
+## Design
+
+Based on the core [Couler APIs design](https://github.com/couler-proj/couler/pull/24), we can see the definitions in different workflow engines.
+
+| Couler | Argo | Tekton |
+| ---- | ---- | ---- |
+| Step | Step | Step |
+| Reusable step | Template | Task |
+| Workflow | Workflow | Pipeline |
+
+* `Step` is the smallest unit and remain consistent across different backends so the core operation `run_step(step_def)` will also keep consistent. In tekton, the definition of `step` and `container` keeps the same, which means that if users pass a python function to step, the backend should wrap the function in a python base image automatically.
+
+### Control Flow
+
+#### `when(cond, if_op, else_op)`
+
+In the latest v0.16 version of Tekton, the condition field is deprecated and Tekton uses [whenExpression](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#guard-task-execution-using-whenexpressions) instead. Here is an example:
+
+```yaml
+tasks:
+  - name: echo-file-exists
+    when:
+      - input: "$(tasks.check-file.results.exists)"
+        operator: in
+        values: ["yes"]
+    taskRef:
+        name: echo-file-exists
+```
+
+A `whenExpression` is made of `Input`, `Operator` and `Values`. The `Input` can be static inputs or variables (Parameters or Results in Tekton). The `Operator` can be either `in` or `notin`.
+As we can see, there is no `else_op` in Tekton, so we need to convert the `else` in the Tekton backend, which is pretty simple since there is only `in` and `notin` operator, we can simply change the operator in the "else" task like:
+```yaml
+tasks:
+  - name: echo-file-not-exists
+    when:
+      - input: "$(tasks.check-file.results.exists)"
+        operator: notin
+        values: ["yes"]
+    taskRef:
+        name: echo-file-not-exists
+```
+
+
+#### `while_loop(cond, func, *args, **kwargs)`
+
+There is no native support for loops in Tekton for now.
+
+### Couler Utilities:
+
+The `get_status`, `get_logs` remains the same as the Argo design.
+
+* `submit(config=workflow_config(schedule="* * * * 1"))`: There is no native support for cron in Tekton, see this [issue](https://github.com/tektoncd/triggers/issues/69) for more information. The community recommend cronjob in kubernetes, see [example](https://github.com/tektoncd/triggers/blob/master/examples/cron/README.md) here.
+* `delete_workflow(workflow_name)`: To be discussed: should we delete the tasks when delete the pipeline?
+* `list_workflow_records(workflow_name)`: After submitting the workflow(create PipelineRun in Tekton), there will be numbers of workflow records. A list function is necessary for users to show all the records.
```

Comment:
Can the name be simplified to `list_workflows()`? We can probably add this in core Couler APIs as well.

Reply:
Also do we have API for dealing with workflow runs? For argo is that a feature only accessible via argo serve API?

Reply:
@binarycrayon Could you clarify what you meant by "dealing with workflow runs"?

Reply:
By mentioning workflow records, I mean `workflow runs`, which is `pipeline run` in tekton. Should we change the name to `list_workflow_runs` and add another api like `list_workflows`?

Reply:
`PipelineRun` is just an instantiation of a Tekton `Pipeline`. In Argo, there's no such additional concept. Should we abstract such backend specific details away so that users will only focus on defining `Workflow` that will be submitted and won't be confused with these two concepts?

Reply:
@terrytangyuan We could, I can add argo server API support as part of `argo-client-python`

Reply:
I think we can keep both `list_workflows` and `list_workflow_runs`. Though this is less applicable to Argo, it's quite common for other backends - they have the concept of `pipeline/workflow` to represent the execution logic and `pipeline/workflow runs` to represent the result and state for each run either triggered by cron or on-demand.

For Argo, I think `CronWorkflow` also has the idea of (though not explicit) the `workflow` vs `run`.

Reply:
Actually for Tekton users, my understanding is that they may not need to get a list of `Pipeline`s at all. Instead, they want to get a list of `PipelineRun`s, which is actually "equivalent" to Argo's `Workflow`. Am I right? If so, we actually only need one API which can be named either `list_workflow_runs` or `list_workflows` where `list_workflows` might be more generalized and concise. Correct me if I am wrong though.